### PR TITLE
bearer: Improvements and test coverage

### PR DESCRIPTION
**Pull request by cthulhu-rider**
* after #612

### CODE REVIEW

File: bearer/bearer.go [hunk 1]
```diff
@@ -204,7 +216,7 @@ func (b *Token) SetIat(iat uint64) {
 //
 // See also SetExp, SetNbf, SetIat.
 func (b Token) InvalidAt(epoch uint64) bool {
-	return !b.lifetimeSet || b.nbf > epoch || b.iat > epoch || b.exp < epoch
+	return b.Nbf() > epoch || b.Iat() > epoch || b.Exp() < epoch
 }
```
Comment (smallhive) [inline bearer/bearer.go:220]: I suggest renaming this method to `ValidAt` and inverse all conditions. I wouldn't say I like methods and variables with negative meanings in names. It is intricate to use them in code. I had experience with this method not so long time ago - it wasn't easy enough

We may mark the current as obsolete, then add a new method

File: bearer/bearer_test.go [hunk 2]
```diff
@@ -214,235 +287,673 @@ func TestToken_InvalidAt(t *testing.T) {
...
-	require.True(t, val.AssertUser(usr))
-	require.False(t, val.AssertUser(usr2))
-
-	var s neofscrypto.Signature
-
-	require.NoError(t, s.CalculateMarshalled(neofscryptotest.Signer(), &body, nil))
-
-	s.WriteToV2(&sig)
-
+	var val bearer.Token
```
Comment (smallhive) [inline bearer/bearer_test.go:580]: Could you please reorganise this code? Add some blank lines for instance. It is difficult to read it.

The same for some other places

File: bearer/bearer.go (hunk 1 above)
Comment (cthulhu-rider) [reply bearer/bearer.go]: done

File: bearer/bearer_test.go (hunk 2 above)
Comment (cthulhu-rider) [reply bearer/bearer_test.go]: added \n

File: bearer/bearer_test.go [hunk 3]
```diff
@@ -165,235 +287,691 @@ func TestToken_InvalidAt(t *testing.T) {
...
+		size = 3 // byte + uint16
+	} else {
+		size = 5 // byte + uint32
+	}
+	return int(size)
+}
 
-	val.ForUser(usr)
-	require.True(t, val.AssertUser(usr))
+func putVarUint(data []byte, val uint64) int {
```
Comment (roman-khimov) [inline bearer/bearer_test.go:334]: https://pkg.go.dev/github.com/nspcc-dev/neo-go/pkg/io#PutVarUint

File: bearer/bearer_test.go [hunk 4]
```diff
@@ -165,235 +287,691 @@ func TestToken_InvalidAt(t *testing.T) {
...
-
-	token.SetIssuer(usr)
-
-	require.True(t, usr == token.Issuer())
-	require.True(t, usr == token.ResolveIssuer())
+func TestToken_MarshalJSON(t *testing.T) {
+	b, err := json.MarshalIndent(validBearerToken, "", " ")
+	require.NoError(t, err)
+	if string(b) != validJSONBearerToken {
+		// protojson is inconsistent https://github.com/golang/protobuf/issues/1121
```
Comment (roman-khimov) [inline bearer/bearer_test.go:863]: https://pkg.go.dev/github.com/stretchr/testify/require#JSONEq

File: bearer/bearer.go [hunk 5]
```diff
@@ -233,9 +233,18 @@ func (b *Token) SetEACLTable(table eacl.Table) {
...
+// eACL is unset.
+func (b Token) GetEACLTable() (eacl.Table, bool) {
+	return b.eaclTable, b.eaclTableSet
+}
+
 // EACLTable returns extended ACL table set by SetEACLTable.
 //
 // Zero Token has zero eacl.Table.
+//
+// Deprecated: zero property of [eacl.Table] is unrecognizable, use
```
Comment (roman-khimov) [inline bearer/bearer.go:251]: It's much easier to _always_ have some table. Can we have a token now without it? Not a real one for sure.

Comment (cthulhu-rider) [reply bearer/bearer.go:251]: > Can we have a token now without it?

there is not cases allowed by the protocol. But may be in some stages of the user app: zero instance, after `Unmarshal`, etc.

Comment (roman-khimov) [reply bearer/bearer.go:251]: Well, then zero instance should have an empty table.

File: bearer/bearer_test.go (hunk 4 above)
Comment (cthulhu-rider) [reply bearer/bearer_test.go:863]: dont see how it can help me here

File: bearer/bearer_test.go [hunk 6]
```diff
@@ -165,235 +287,691 @@ func TestToken_InvalidAt(t *testing.T) {
...
+				j: `{"body":{"issuer":{}}}`},
+			{name: "body/issuer/value/undersize", err: "invalid issuer: invalid length 24, expected 25",
+				j: `{"body":{"issuer":{"value":"NTMFpm8dFGXApRynOaBSUCnLFP4eisMR"}}}`},
+			{name: "body/issuer/value/oversize", err: "invalid issuer: invalid length 26, expected 25",
+				j: `{"body":{"issuer":{"value":"NTMFpm8dFGXApRynOaBSUCnLFP4eisMRXAE="}}}`},
+			{name: "body/issuer/value/wrong prefix", err: "invalid issuer: invalid prefix byte 0x42, expected 0x35",
+				j: `{"body":{"issuer":{"value":"QjMFpm8dFGXApRynOaBSUCnLFP4eisMRXA=="}}}`},
+			{name: "body/issuer/value/checksum mismatch", err: "invalid issuer: checksum mismatch",
+				j: `{"body":{"issuer":{"value":"NTMFpm8dFGXApRynOaBSUCnLFP4eisMRXQ=="}}}`},
+			// {name: "signature/invalid scheme", err: "invalid body signature: scheme 2147483648 overflows int32",
```
[the remark this answers is not in the document]
Comment (cthulhu-rider) [reply bearer/bearer_test.go:905]: nein

File: bearer/bearer.go (hunk 5 above)
Comment (cthulhu-rider) [reply bearer/bearer.go:251]: it has, but we cannot determine table emptiness

File: bearer/bearer_test.go (hunk 4 above)
Comment (cthulhu-rider) [reply bearer/bearer_test.go]: done

File: bearer/bearer_test.go [hunk 7]
```diff
@@ -165,235 +287,691 @@ func TestToken_InvalidAt(t *testing.T) {
...
+	n := copy(b, []byte{0x01, 0x00, 0x01, 0xf0})
+	n += putVarUint(b[n:], uint64(payloadLen))
+	n += hex.Encode(b[n:], sig[64:])
+	n += copy(b[n:], b64)
+	copy(b[n:], []byte{0x00, 0x00})
+	h256 := sha256.Sum256(b)
+	r, s = new(big.Int).SetBytes(sig[:32]), new(big.Int).SetBytes(sig[32:][:32])
+	require.True(t, ecdsa.Verify(&ecdsaPriv.PublicKey, h256[:], r, s))
+
+	/* determenistic schemes */
```
[the remark this answers is not in the document]
Comment (cthulhu-rider) [reply bearer/bearer_test.go]: fixed

**Review by smallhive — APPROVED** [review]

File: bearer/bearer.go (hunk 5 above)
Comment (roman-khimov) [reply bearer/bearer.go:251]: Why do you care about emptiness? What's wrong with the definition of empty table from https://pkg.go.dev/github.com/nspcc-dev/neofs-sdk-go/eacl#NewTable?

**Review by roman-khimov — COMMENTED** [review]
Looks like 2a37283cd802a40a54d1c106e6bac5382e8f3225 does much more than expected from its message.

File: bearer/bearer.go (hunk 5 above)
Comment (cthulhu-rider) [reply bearer/bearer.go:251]: i dont care about table emptiness, i care about the presence which should be determined somehow. Boolean flag is the one-condition testable way. Other one could be pointer return which i dont like

**Comment by cthulhu-rider** [comment]
oops rebase sheet, fixed

File: bearer/bearer.go (hunk 5 above)
Comment (roman-khimov) [reply bearer/bearer.go:251]: For what purpose? All proper tokens have tables. Token without a table doesn't exists. Can we make it like that?

Comment (cthulhu-rider) [reply bearer/bearer.go:251]: blank instance with unset table

Comment (roman-khimov) [reply bearer/bearer.go:251]: > blank instance with unset table

Blank instances should have an empty table. Don't allow to create it and problem solved. SDK interfaces should be easy to use.

Comment (cthulhu-rider) [reply bearer/bearer.go:251]: thats it, there is **no** easy interface to check table for zero (incl. absence). This could be `IsZero` interface which could replace proposed method. https://pkg.go.dev/github.com/nspcc-dev/neofs-sdk-go/eacl#NewTable does not fit cuz it returns particular non-zero table and checking field-by-field isnt be forward compatible

Comment (roman-khimov) [reply bearer/bearer.go:251]: I don't understand the problem. Look at how `EACLTable()` is used by the node, for example. All of these cases will do `t, _ = GetEACLTable()`. No code cares about this bool. No code should care.

Comment (roman-khimov) [reply bearer/bearer.go:251]: You can drop `eaclTableSet` altogether. No one would notice.

Comment (cthulhu-rider) [reply bearer/bearer.go:251]: > No code should care.

table can be omitted in the just created instance or in the protocol message. This case should be observable. The fact that some code may rely on indirect signs of absence, such as the omission of a signature, does not mean that no one cares whether the table is specified. I'd like to see the case when smbd has completely forgotten the table, although this can be drawn to a more specific "there is something wrong with the token...". Makes easier to understand the protocol violations

> You can drop eaclTableSet altogether. No one would notice.

i know, it's private

Comment (roman-khimov) [reply bearer/bearer.go:251]: > just created instance

Should always have it.

> in the protocol message

Decoding error (that we have already anyway).

Comment (cthulhu-rider) [reply bearer/bearer.go:251]: > Should always have it.

there is no way to ensure this

Comment (carpawell) [reply bearer/bearer.go:251]: what code except unmarshalers needs this information? `ReadFromV2` just requires it

Comment (cthulhu-rider) [reply bearer/bearer.go:251]: API client ops may require parameterized token validation

Comment (carpawell) [reply bearer/bearer.go:251]: API client always requires v2 version inside, right? my main concern is usability. i feel like there will be no one (or literally a single strange case) `ok` flag usage in practice but also everybody will be forced to suppress this flag

it was always a pain in some part of me to use skip flags like in https://github.com/nspcc-dev/neofs-sdk-go/blob/3a8e90ffe366f0f562878c830de8a700265736cf/object/object.go#L246 and now we have https://github.com/nspcc-dev/neofs-sdk-go/blob/3a8e90ffe366f0f562878c830de8a700265736cf/object/object.go#L267

years of suffering to understand that it was not necessary. nothing changed in fact, we just have admitted that zero ID is not possible _in practice_ and added some words somewhere. the same way i do not see what user code _in practice_ will use it. even if it is required, thousands of `table, _` and a single usecase are worse to me than a separate method added after we have found it is required by something

Comment (cthulhu-rider) [reply bearer/bearer.go:251]: yeah, zero prop is very good. Note that it's stated even [now](https://pkg.go.dev/github.com/nspcc-dev/neofs-sdk-go/bearer#Token.EACLTable). But then the question is - how to check? Zero IDs are easy to detect and we added dedicated [method](https://github.com/nspcc-dev/neofs-sdk-go/blob/3a8e90ffe366f0f562878c830de8a700265736cf/container/id/id.go#L153). Moreover, ID types are comparable and can be zerochecked via `==`. But most of the structs like eACL cannot

it should be noted that zeroness removes the diff b/w a complete absence of a field and a zero but specified one. Such a difference can only be emphasized by an additional flag. Not a eACL case

so, as i said above, im all in for returning struct only, but with exposed `IsZero` method. The code would be:
```go
eACL := token.EACLTable()
if eACL.IsZero() { // only if i care
  // fail
}
// other stuff
```

---

to sum up, i see next options:
1. change nothing. Always seems nicer, but the fact of the table absence will continue to merge with indirect problems of "there is no such or such field"
2. as proposed now (mb with dropped deprecation mark and keeping both methods and letting user to decide what to use)
3. add `eacl.Table.IsZero` method and leave token untouched. Reuse ID approach and, again, let user decide to check or not to check. It will be very leghtweight btw

my choice is 3

Comment (carpawell) [reply bearer/bearer.go:251]: ok with 3 too. @roman-khimov

Comment (roman-khimov) [reply bearer/bearer.go:251]: OK for `IsZero`.

Comment (cthulhu-rider) [reply bearer/bearer.go]: done

File: eacl/table.go [hunk 8]
```diff
@@ -274,3 +274,9 @@ func (t *Table) UnmarshalJSON(data []byte) error {
 // EqualTables compares Table with each other.
 // Deprecated: compare [Table.Marshal] instead.
 func EqualTables(t1, t2 Table) bool { return bytes.Equal(t1.Marshal(), t2.Marshal()) }
+
+// IsZero checks whether all fields of the table are zero/empty. The property
+// can be used as a marker of unset eACL.
+func (t Table) IsZero() bool {
+	return t.cid.IsZero() && len(t.records) == 0 && t.version == version.Version{}
```
Comment (roman-khimov) [inline eacl/table.go:281]: Not covered.

**Review by roman-khimov — APPROVED** [review]

**Review by carpawell — APPROVED** [review]
